src/test/java/com/balako/onlinebookstore/service/BookServiceTest.java

```diff
@@ -0,0 +1,243 @@
+    private static final BookDto BOOK_DTO = new BookDto();
+    private static final Long BOOK_ID = 1L;
+    private static final Long CATEGORY_ID = 1L;
+    private static final Integer NUMBER_OF_INVOCATIONS = 1;
+    private static final BookSearchParametersDto SEARCH_PARAMETERS =
+            new BookSearchParametersDto(
+                    new String[] {"Title 1", "Title 2"},
+                    new String[] {"Author 1", "Author 2"},
+                    new String[] {"978-0545010221"});
+    @InjectMocks
+    private BookServiceImpl bookService;
+    @Mock
+    private BookRepository bookRepository;
+    @Mock
+    private BookMapper bookMapper;
+    @Mock
+    private BookSpecificationBuilder bookSpecificationBuilder;
+
+    @BeforeAll
+    static void beforeAll() {
+        CATEGORY.setId(CATEGORY_ID);
+        CATEGORY.setName("Category 1");
+        CATEGORY.setDescription("Category 1 description");
+
+        BOOK.setId(BOOK_ID);
+        BOOK.setTitle("Title 1");
+        BOOK.setAuthor("Author 1");
+        BOOK.setIsbn("978-0545010221");
+        BOOK.setDescription("Some description");
+        BOOK.setCoverImage("Image 1");
+        BOOK.setCategories(Set.of(CATEGORY));
+        BOOK.setPrice(BigDecimal.TEN);
+
+        REQUEST_DTO.setTitle(BOOK.getTitle());
+        REQUEST_DTO.setAuthor(BOOK.getAuthor());
+        REQUEST_DTO.setIsbn(BOOK.getIsbn());
+        REQUEST_DTO.setDescription(BOOK.getDescription());
+        REQUEST_DTO.setCoverImage(BOOK.getCoverImage());
+        REQUEST_DTO.setPrice(BOOK.getPrice());
+        REQUEST_DTO.setCategoryIds(Set.of(CATEGORY.getId()));
+
+        BOOK_DTO.setId(BOOK.getId());
+        BOOK_DTO.setTitle(BOOK.getTitle());
+        BOOK_DTO.setAuthor(BOOK.getAuthor());
+        BOOK_DTO.setIsbn(BOOK.getIsbn());
+        BOOK_DTO.setDescription(BOOK.getDescription());
+        BOOK_DTO.setCoverImage(BOOK.getCoverImage());
+        BOOK_DTO.setPrice(BOOK.getPrice());
+        BOOK_DTO.setCategories(REQUEST_DTO.getCategoryIds());
+
+        DTO_WITHOUT_CATEGORY_IDS.setId(BOOK.getId());
+        DTO_WITHOUT_CATEGORY_IDS.setTitle(BOOK.getTitle());
+        DTO_WITHOUT_CATEGORY_IDS.setAuthor(BOOK.getAuthor());
+        DTO_WITHOUT_CATEGORY_IDS.setIsbn(BOOK.getIsbn());
+        DTO_WITHOUT_CATEGORY_IDS.setDescription(BOOK.getDescription());
+        DTO_WITHOUT_CATEGORY_IDS.setCoverImage(BOOK.getCoverImage());
+        DTO_WITHOUT_CATEGORY_IDS.setPrice(BOOK.getPrice());
+    }
+
+    @Test
+    @DisplayName("Test saving a book with valid request")
+    public void save_validBook_returnBookDto() {
+        when(bookRepository.save(any(Book.class))).thenReturn(BOOK);
+        when(bookMapper.toDto(any(Book.class))).thenReturn(BOOK_DTO);
+        when(bookMapper.toModel(any(CreateBookRequestDto.class)))
+                .thenReturn(BOOK);
+
+        BookDto actual = bookService.save(REQUEST_DTO);
+
+        assertEquals(BOOK_DTO, actual);
+        verify(bookRepository, times(NUMBER_OF_INVOCATIONS))
```

Comment:
calling `times` is redundant when you want to validate that it was called 1 time, because it is a default behaviour.
![image](https://github.com/Denis-Balako/Online-Book-Store/assets/82539553/6f1fc4d3-4f7f-4a27-ad67-697c2a29ce5e)